README


This archive contains the distribution QBit-WebInterface-Routing,
version 0.008:

  Class for creating routing for web interface

This software is Copyright (c) 2016 by Igor Sverdlov.

This is free software, licensed under:

  The GNU General Public License, Version 2, June 1991


This README file was generated by Dist::Zilla::Plugin::Readme v5.039.



Changes
==================================================
Changes from 2015-12-21 00:00:00 +0000 to present.
==================================================

------------------------------------------
version 0.008 at 2016-12-20 05:03:41 +0000
------------------------------------------

  Change: c72cf3519e9f6f62e6828739a7a7c3f4f9dc3561
  Author: Igor <igor-sverdlov@mail.ru>
  Date : 2016-12-20 08:03:03 +0000

    Merge pull request #10 from QBitFramework/exceptions

    exceptions 

  Change: 8dc9d46411dac82474afd5d2b70d36d90ef429f4
  Author: Igor Sverdlov <igor-sverdlov@mail.ru>
  Date : 2016-12-20 08:01:33 +0000

    exceptions 

------------------------------------------
version 0.007 at 2016-12-04 16:59:53 +0000
------------------------------------------

  Change: 8f71832e3dec8fbe495cedca9b6991062c54eb52
  Author: Igor Sverdlov <igor-sverdlov@mail.ru>
  Date : 2016-12-04 19:59:53 +0000

    Version 0.007 

  Change: ac1ab5a0c3da253045691bda2eceb97bd63005da
  Author: Igor <igor-sverdlov@mail.ru>
  Date : 2016-12-04 19:58:48 +0000

    Merge pull request #9 from QBitFramework/same_routing

    two routes with difference methods 

  Change: 2636366b576d0b14fdc3d6a1770ba2c2830cbb94
  Author: Igor Sverdlov <igor-sverdlov@mail.ru>
  Date : 2016-12-04 19:07:15 +0000

    two routes with difference methods 

------------------------------------------
version 0.006 at 2016-10-09 22:33:36 +0000
------------------------------------------

  Change: 4c46f19f97bf6485a2c2eef56c31e1cf77d3a3bf
  Author: Igor Sverdlov <igor-sverdlov@mail.ru>
  Date : 2016-10-10 01:33:36 +0000

    Version 0.006 

  Change: 7366d28abb263b056d2ab1ee406b52658f6c4de6
  Author: Igor <igor-sverdlov@mail.ru>
  Date : 2016-10-10 02:31:55 +0000

    Merge pull request #8 from QBitFramework/fix_dependency

    fix dependency 

  Change: ec5400f0b1dde24264933aaf9dbd058e3dba216d
  Author: Igor Sverdlov <igor-sverdlov@mail.ru>
  Date : 2016-10-10 01:29:28 +0000

    fix dependency 

------------------------------------------
version 0.005 at 2016-10-09 15:34:07 +0000
------------------------------------------

  Change: a1a42414c54241376f3c2b0efa489e5ef229ea7b
  Author: Igor Sverdlov <igor-sverdlov@mail.ru>
  Date : 2016-10-09 18:34:07 +0000

    Version 0.005 

  Change: 2ce86327aa1154ebbb81ec08f14cc95ea7da41c5
  Author: Igor <igor-sverdlov@mail.ru>
  Date : 2016-10-09 19:32:56 +0000

    Merge pull request #7 from QBitFramework/change_placeholders

    fix README.md 

  Change: 9281794f1c62e2f0d1bfcded5a4cf2880277d8de
  Author: Igor Sverdlov <igor-sverdlov@mail.ru>
  Date : 2016-10-09 18:28:48 +0000

    fix README.md 

------------------------------------------
version 0.004 at 2016-10-09 15:26:49 +0000
------------------------------------------

  Change: fe915c519460485c21712450879c7baae1932acd
  Author: Igor Sverdlov <igor-sverdlov@mail.ru>
  Date : 2016-10-09 18:26:49 +0000

    Version 0.004 

  Change: 47f77ca088db6b836acf44e363f9504021702f1e
  Author: Igor <igor-sverdlov@mail.ru>
  Date : 2016-10-09 19:25:24 +0000

    Merge pull request #6 from QBitFramework/change_placeholders

    change placeholders : -> !, # -> : 

  Change: 6e12737ed20697b795e9bf90e19d3ef6ddf2747d
  Author: Igor Sverdlov <igor-sverdlov@mail.ru>
  Date : 2016-10-09 17:36:18 +0000

    change placeholders : -> !, # -> : 

------------------------------------------
version 0.003 at 2016-10-08 19:48:07 +0000
------------------------------------------

  Change: f6c13dc4c1a881b2b31fbf1eddb92d449a1962f3
  Author: Igor Sverdlov <igor-sverdlov@mail.ru>
  Date : 2016-10-08 22:48:07 +0000

    Version 0.003 

  Change: c6f3674d74b146d1561ede03adff0017264ecd40
  Author: Igor <igor-sverdlov@mail.ru>
  Date : 2016-10-08 23:47:24 +0000

    Merge pull request #5 from QBitFramework/move_t_lib

    t_lib -> t/lib 

  Change: 51f053194962986575012e60642b8b154c11f089
  Author: Igor Sverdlov <igor-sverdlov@mail.ru>
  Date : 2016-10-08 22:45:11 +0000

    t_lib -> t/lib 

------------------------------------------
version 0.002 at 2016-10-08 13:43:12 +0000
------------------------------------------

  Change: 15959a177121506e48db3b839febd5eaca6ff27f
  Author: Igor Sverdlov <igor-sverdlov@mail.ru>
  Date : 2016-10-08 16:43:12 +0000

    Version 0.002 

  Change: c5334ad66eb6fe960ef9e7e442a31bd93d122a5c
  Author: Igor <igor-sverdlov@mail.ru>
  Date : 2016-10-08 17:41:23 +0000

    Merge pull request #4 from QBitFramework/fix_dependency

    TestWebInterface -> TWebInterface 

  Change: 28ff0fdac1c19344830cdc3338992576509412c3
  Author: Igor Sverdlov <igor-sverdlov@mail.ru>
  Date : 2016-10-08 16:39:18 +0000

    TestWebInterface -> TWebInterface 

------------------------------------------
version 0.001 at 2016-09-11 14:50:33 +0000
------------------------------------------

  Change: b3e1f835212868c6258af017c2d814d6eb063d95
  Author: Igor Sverdlov <igor-sverdlov@mail.ru>
  Date : 2016-09-11 17:50:33 +0000

    Version 0.001 

  Change: f8a4c423c3a8b99e3e5ff7cedad4f2a9e6c9946b
  Author: Igor <igor-sverdlov@mail.ru>
  Date : 2016-09-11 18:07:25 +0000

    Merge pull request #1 from QBitFramework/upload_to_cpan

    upload to cpan 

  Change: a3e120f1eef70b07395d4b42d4d9e2c34081cc1a
  Author: Igor Sverdlov <igor-sverdlov@mail.ru>
  Date : 2016-09-11 17:06:16 +0000

    upload to cpan 

  Change: 11d6d1ea63a35d0e0b61fd6efc1e9eebc0b1f2b3
  Author: Igor Sverdlov <igor-sverdlov@mail.ru>
  Date : 2016-02-14 20:36:15 +0000

    fix README 

  Change: 773cbb222adb991ea7a4d64e628bc53d86690ffc
  Author: Igor Sverdlov <igor-sverdlov@mail.ru>
  Date : 2016-02-14 20:24:25 +0000

    package - ok 

  Change: 22d7117c9312bebdfb7890176a70721e5deda6ef
  Author: Igor Sverdlov <igor-sverdlov@mail.ru>
  Date : 2016-02-11 00:44:07 +0000

    fix 

  Change: e0620108362a78eb81e81d2b62dfa2b1a17f8475
  Author: blizzard <blizzard@yandex-team.ru>
  Date : 2016-02-10 20:21:36 +0000

    fix utf8 

  Change: 5acace75a37430041e77c572b3889655c57a6c24
  Author: blizzard <blizzard@yandex-team.ru>

Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.039.
use strict;
use warnings;



use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Class for creating routing for web interface",
  "AUTHOR" => "Igor Sverdlov <igor-sverdlov\@mail.ru>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "QBit-WebInterface-Routing",
  "LICENSE" => "gpl",
  "NAME" => "QBit::WebInterface::Routing",
  "PREREQ_PM" => {
    "Exception" => 0,
    "File::Path" => 0,
    "POSIX" => 0,
    "QBit::Application" => 0,
    "QBit::Class" => 0,
    "QBit::WebInterface" => 0,
    "QBit::WebInterface::Response" => 0,
    "URI::Escape" => 0,
    "base" => 0,
    "qbit" => 0,
    "strict" => 0
  },
  "TEST_REQUIRES" => {
    "Digest::MD5" => 0,
    "QBit::WebInterface::Controller" => 0,
    "QBit::WebInterface::Request" => 0,
    "Test::Deep" => 0,
    "Test::More" => 0,
    "lib::abs" => 0,
    "warnings" => 0
  },
  "VERSION" => "0.008",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Digest::MD5" => 0,
  "Exception" => 0,
  "File::Path" => 0,
  "POSIX" => 0,
  "QBit::Application" => 0,
  "QBit::Class" => 0,
  "QBit::WebInterface" => 0,
  "QBit::WebInterface::Controller" => 0,
  "QBit::WebInterface::Request" => 0,
  "QBit::WebInterface::Response" => 0,
  "Test::Deep" => 0,
  "Test::More" => 0,
  "URI::Escape" => 0,
  "base" => 0,
  "lib::abs" => 0,
  "qbit" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);


lib/QBit/WebInterface/Routing.pm
package QBit::WebInterface::Routing;

use POSIX qw(strftime setlocale LC_TIME);

use qbit;

use QBit::WebInterface::Routing::Routes;
use QBit::WebInterface::Response;

use base qw(QBit::WebInterface QBit::Application);

sub import {
    my ($package, %opts) = @_;

    $package->SUPER::import(%opts);

    my $package_wi = caller(0);

    die gettext('Use only in QBit::WebInterface and QBit::Application descendant')
      unless $package_wi->isa('QBit::WebInterface')
          && $package_wi->isa('QBit::Application');

    {
        no strict 'refs';

        *{"${package_wi}::build_response"}     = \&build_response;
        *{"${package_wi}::routing"}            = \&routing;
        *{"${package_wi}::exception_handling"} = \&exception_handling;
    }
}

sub routing {
    my ($self, %opts) = @_;

    $self->{'__ROUTING__'} = QBit::WebInterface::Routing::Routes->new(%opts);

    return $self->{'__ROUTING__'};
}

sub build_response {
    my ($self) = @_;

    $self->pre_run();

    throw gettext('No request object') unless $self->request;
    $self->response(QBit::WebInterface::Response->new());

    my $cmds = $self->get_cmds();

    my $path   = '';
    my $cmd    = '';
    my %params = ();

    if ($self->{'__ROUTING__'}) {
        my $route = $self->{'__ROUTING__'}->get_current_route($self);

        if (exists($route->{'handler'})) {
            $path = '__HANDLER_PATH__';
            $cmd  = '__HANDLER_CMD__';

            my $package = $self->get_option('controller_class', 'QBit::WebInterface::Controller');

            my $imported = $self->{'__IMPORTED__'}{$package};

            foreach my $p (keys(%$cmds)) {
                last if $imported;

                foreach my $c (keys(%{$cmds->{$p}})) {
                    if ($cmds->{$p}{$c}{'package'} eq $package) {
                        $imported = TRUE;

                        last;
                    }
                }
            }

            unless ($imported) {
                my $req_package = $package . '.pm';
                $req_package =~ s/::/\//g;

                require $req_package;

                $package->import();

                $self->{'__IMPORTED__'}{$package} = TRUE;
            }

            $cmds->{$path}{$cmd} = {
                'package'        => $package,
                'sub'            => $route->{'handler'},
                'type'           => $route->{'type'} // 'CMD',
                'process_method' => $route->{'process_method'},
                'attrs'          => {map {$_ => TRUE} @{$route->{'attrs'} // []}}
            };
        } else {
            $path = $route->{'path'} // '';
            $cmd  = $route->{'cmd'}  // '';
        }

        %params = %{$route->{'args'} // {}};
    }

    if ($self->get_option('use_base_routing') && !(length($path) || length($cmd))) {
        ($path, $cmd) = $self->get_cmd();

        $cmd = $cmds->{$path}{'__DEFAULT__'}{'name'} if $cmd eq '';
        $cmd = '' unless defined($cmd);
    }

    $self->set_option(cur_cmd     => $cmd);
    $self->set_option(cur_cmdpath => $path);

    if ($self->{'__EXCEPTION_IN_ROUTING__'}) {
        #nothing do...
    } elsif (exists($cmds->{$path}{$cmd})) {
        try {
            my $cmd = $cmds->{$path}{$cmd};

            my $controller = $cmd->{'package'}->new(
                app   => $self,
                path  => $path,
                attrs => $cmd->{'attrs'}
            );

            $self->{'__BREAK_PROCESS__'} = 0;
            $self->pre_cmd($controller);

            unless ($self->{'__BREAK_PROCESS__'}) {
                $controller->{'__BREAK_CMD__'} = FALSE;
                $controller->pre_cmd() if $controller->can('pre_cmd');

                unless ($controller->{'__BREAK_CMD__'}) {
                    if ($controller->attrs()->{'SAFE'}) {
                        throw Exception::WebInterface::Controller::CSRF gettext('CSRF has been detected')
                          unless $controller->check_anti_csrf_token(
                            $self->request->param(sign => $params{'sign'} // ''),
                            url => $self->get_option('cur_cmdpath') . '/' . $self->get_option('cur_cmd'));
                    }

                    my @data = $cmd->{'sub'}($controller, %params);
                    if (defined(my $method = $cmd->{'process_method'})) {
                        $controller->$method(@data);
                    }
                }
            }

            $self->post_cmd();
        }
        catch Exception::Denied with {
            $self->response->status(403);
            $self->response->data(undef);
        }
        catch Exception::Request::UnknownMethod with {
            $self->response->status(400);
            $self->response->data(undef);
        }
        catch {
            $self->exception_handling(shift);
        };
    } else {
        $self->response->status(404);
    }

    my $ua = $self->request->http_header('User-Agent');
    $self->response->headers->{'Pragma'} = ($ua =~ /MSIE/) ? 'public' : 'no-cache';

    $self->response->headers->{'Cache-Control'} =
      ($ua =~ /MSIE/)
      ? 'must-revalidate, post-check=0, pre-check=0'
      : 'no-cache, no-store, max-age=0, must-revalidate';

    my $tm   = time();
    my $zone = (strftime("%z", localtime($tm)) + 0) / 100;
    my $loc  = setlocale(LC_TIME);
    setlocale(LC_TIME, 'en_US.UTF-8');
    my $GMT = strftime("%a, %d %b %Y %H:%M:%S GMT", localtime($tm - $zone * 3600));
    setlocale(LC_TIME, $loc);

    $self->response->headers->{'Expires'} = $GMT;

    $self->post_run();

    $self->response->timelog($self->timelog);
}

sub exception_handling {
    my ($self, $exception) = @_;

    if (my $dir = $self->get_option('error_dump_dir')) {
        require File::Path;
        File::Path::make_path($dir);
        writefile("$dir/dump_" . format_date(curdate(), '%Y%m%d_%H%M%S') . "${$}.html",
            $self->_exception2html($exception));
        $self->response->status(500);
        $self->response->data(undef);
    } else {
        if (($self->request->http_header('Accept') || '') =~ /(application\/json|text\/javascript)/) {
            $self->response->content_type("$1; charset=UTF-8");
            $self->response->data(to_json({error => $exception->message()}));
        } else {
            $self->response->data($self->_exception2html($exception));
        }
    }
}

TRUE;

__END__

=encoding utf8

=head1 Name

QBit::WebInterface::Routing - Class for creating routing for web interface.

=head1 GitHub

https://github.com/QBitFramework/QBit-WebInterface-Routing

=head1 Install

=over

=item *

cpanm QBit::WebInterface::Routing

=item *

apt-get install libqbit-webinterface-routing-perl (http://perlhub.ru/)

=back

For more information. please, see code.

=cut


lib/QBit/WebInterface/Routing/Routes.pm
package QBit::WebInterface::Routing::Routes;

use qbit;

use base qw(QBit::Class);

use Exception::Routes;

use URI::Escape qw(uri_escape_utf8 uri_unescape);

my %METHODS = (
    GET     => 1,
    HEAD    => 2,
    POST    => 4,
    PUT     => 8,
    PATCH   => 16,
    DELETE  => 32,
    OPTIONS => 64,
);

sub init {$_[0]->{'strictly'} //= 1;}

sub get {shift->_generate_route(GET => @_)}

sub head {shift->_generate_route(HEAD => @_)}

sub post {shift->_generate_route(POST => @_)}

sub put {shift->_generate_route(PUT => @_)}

sub patch {shift->_generate_route(PATCH => @_)}

sub delete {shift->_generate_route(DELETE => @_)}

sub options {shift->_generate_route(OPTIONS => @_)}

sub any {
    my ($self, @args) = @_;

    if (@args == 1) {
        shift->_generate_route([keys(%METHODS)] => @args);
    } elsif (@args == 2) {
        shift->_generate_route(@args);
    } else {
        throw Exception::Routes gettext('Expected one or two arguments');
    }
}

sub under {
    my ($self, @args) = @_;

    throw Exception::Routes gettext('Route must begin with "/"') unless $args[0] =~ m/\A\//;

    $args[0] =~ s/\/\z//;

    my $under_route = $self->new(strictly => $self->{'strictly'});

    $under_route->{'__UNDER__'} = {route => $args[0]};

    $under_route->{'__LAST__'} = \$under_route->{'__UNDER__'};

    $under_route->{'__ROUTES__'} = $self->{'__ROUTES__'};

    return $under_route;
}

sub _generate_route {
    my ($self, $methods, $url) = @_;

    throw Exception::Routes gettext('Route must begin with "/"') unless $url =~ m/\A\//;

    if (exists($self->{'__UNDER__'})) {
        $url =~ s/\A\/// if $self->{'__UNDER__'}{'route'} =~ m/\/\z/;

        $url = $self->{'__UNDER__'}{'route'} . $url;
    }

    my $methods_bits = $self->_get_methods_bit($methods);

    my $route = {
        methods => $methods_bits,
        %{$self->_get_settings($url)}
    };

    if (exists($self->{'__UNDER__'})) {
        $route->{'route_path'} = $self->{'__UNDER__'}{'route_path'}
          if exists($self->{'__UNDER__'}{'route_path'});

        $route->{'conditions'} = $self->{'__UNDER__'}{'conditions'}
          if exists($self->{'__UNDER__'}{'conditions'});
    }

    foreach my $r (@{$self->{'__ROUTES__'} // []}) {
        if ($route->{'route_name'} eq $r->{'route_name'} && $route->{'methods'} & $r->{'methods'}) {
            throw Exception::Routes gettext('Route "%s" already exists', $url);
        }
    }

    push(@{$self->{'__ROUTES__'}}, $route);

    $self->{'__LAST__'} = \$route;

    return $self;
}

sub _get_settings {
    my ($self, $route_name) = @_;

    my @route_levels = split('/', $route_name);

    my @params        = ();
    my @format_levels = ();
    foreach my $route_level (@route_levels) {
        my $regexp_and_format = $self->_get_regexp_and_format_for_level($route_level, \@params);
        $route_level = $regexp_and_format->{'regexp'};
        push(@format_levels, $regexp_and_format->{'format'});
    }

    throw Exception::Routes gettext('Placeholders names can not intersect')
      if @params > 1 && @{arrays_intersection(@params)};

    my $pattern = @route_levels  ? join('\/', @route_levels)  : '\/';
    my $format  = @format_levels ? join('/',  @format_levels) : '/';

    if (!$self->{'strictly'}) {
        $pattern .= '\/' if $pattern !~ m/\/\z/;
        $format  .= '/'  if $format !~ m/\/\z/;
    } elsif ($route_name =~ m/\/\z/) {
        $pattern .= '\/' if $pattern !~ m/\/\z/;
        $format  .= '/'  if $format !~ m/\/\z/;
    }

    return {
        pattern    => '\A' . $pattern . '\z',
        format     => $format,
        params     => \@params,
        levels     => scalar(grep {length($_)} @route_levels),
        route_name => $route_name,
    };
}

sub _get_regexp_and_format_for_level {
    my ($self, $level, $params) = @_;

    my $regexp = '';
    my $format = '';
    my $param  = '';
    my $spec_symbol;
    my $new_spec_symbol;
    my $is_end = FALSE;

    foreach my $symbol (split('', $level)) {
        if (defined($new_spec_symbol)) {
            if ($symbol eq $new_spec_symbol) {
                if (defined($spec_symbol)) {
                    $param .= $symbol;
                } else {
                    $regexp .= quotemeta($symbol);
                    $format .= $symbol;
                }

                undef($new_spec_symbol);

                next;
            } else {
                throw Exception::Routes gettext('You must use "%s" for symbol "%s"', $new_spec_symbol x 2,
                    $new_spec_symbol);
            }
        }

        if ($symbol =~ /[!\:\*]/) {
            if (defined($spec_symbol)) {
                if ($symbol eq $spec_symbol) {
                    if ($is_end) {
                        $is_end = FALSE;

                        $param .= $symbol;
                    } elsif (length($param)) {
                        $is_end = TRUE;
                    } else {
                        undef($spec_symbol);

                        $regexp .= quotemeta($symbol);
                        $format .= $symbol;
                    }
                } elsif ($is_end) {
                    push(@$params, $param);
                    $param = '';

                    $regexp .= $self->_get_regexp_by_symbol($spec_symbol);
                    $format .= '%s';

                    $is_end = FALSE;

                    $spec_symbol = $symbol;
                } else {
                    $new_spec_symbol = $symbol;
                }
            } else {
                $spec_symbol = $symbol;
            }
        } else {
            if ($is_end) {
                push(@$params, $param);
                $param = '';

                $regexp .= $self->_get_regexp_by_symbol($spec_symbol);
                $format .= '%s';

                $is_end = FALSE;

                undef($spec_symbol);

                $regexp .= quotemeta($symbol);
                $format .= $symbol;
            } elsif (defined($spec_symbol)) {
                $param .= $symbol;
            } else {
                $regexp .= quotemeta($symbol);
                $format .= $symbol;
            }
        }
    }

    throw Exception::Routes gettext('You must use "%s" for symbol "%s"', $new_spec_symbol x 2, $new_spec_symbol)
      if $new_spec_symbol;

    if ($is_end) {
        push(@$params, $param);
        $param = '';

        $regexp .= $self->_get_regexp_by_symbol($spec_symbol);
        $format .= '%s';

        undef($spec_symbol);
    }

    throw Exception::Routes gettext('You must use "%s" for symbol "%s"', $spec_symbol x 2, $spec_symbol)
      if $spec_symbol;

    return {regexp => $regexp, format => $format};
}

sub _get_regexp_by_symbol {
    my ($self, $symbol) = @_;

    my $regexp = '';

    if ($symbol eq '!') {
        # /user/!id!
        $regexp = '([^\/.]+)';
    } elsif ($symbol eq ':') {
        # /user/:name:
        $regexp = '([^\/]+)';
    } elsif ($symbol eq '*') {
        # /user/*name*
        $regexp = '(.+)';
    }

    return $regexp;
}

sub _get_methods_bit {
    my ($self, $methods) = @_;

    $methods = [$methods] unless ref($methods) eq 'ARRAY';

    my $methods_bit = $METHODS{shift(@$methods)};
    foreach my $method (@$methods) {
        $methods_bit |= $METHODS{$method};
    }

    return $methods_bit;
}

sub to {
    my ($self, @args) = @_;

    my $route_path;
    if (@args == 2 && $args[0] eq 'controller' && ref($args[1]) eq 'CODE') {
        #to(controller => \&func)

        $route_path = {path => '', cmd => '', controller => $args[1]};
    } elsif (@args % 2 == 0) {
        #to(path => <PATH>, cmd => <CMD>)

        $route_path = {hash_transform({@args}, ['path', 'cmd'])};
    } elsif (@args == 1) {
        if (ref($args[0]) eq 'CODE') {
            #to(\&func)

            $route_path = {path => '', cmd => '', handler => $args[0]};
        } else {
            #to('path#cmd')

            my ($path, $cmd) = ($args[0] =~ m/\A([a-zA-Z_0-9]+)?#([a-zA-Z_][a-zA-Z_0-9]+)?\z/);

            $route_path = {path => $path // '', cmd => $cmd // ''};
        }
    } else {
        throw Exception::Routes gettext('Unknown format arguments');
    }

    if (exists($route_path->{'handler'})) {
        ${$self->{'__LAST__'}}->{'route_path'} = $route_path;
    } else {
        if (exists(${$self->{'__LAST__'}}->{'route_path'})) {
            my $under_route_path = ${$self->{'__LAST__'}}->{'route_path'};

            ${$self->{'__LAST__'}}->{'route_path'} = {
                path       => '',
                cmd        => '',
                controller => sub {
                    my ($web_interface, $params) = @_;

                    if (exists($under_route_path->{'controller'})) {
                        my ($path, $cmd) = $under_route_path->{'controller'}($web_interface, $params);

                        $under_route_path->{'path'} = $path // '';
                        $under_route_path->{'cmd'}  = $cmd  // '';
                    }

                    if (exists($route_path->{'controller'})) {
                        my ($path, $cmd) = $route_path->{'controller'}($web_interface, $params);

                        $route_path->{'path'} = $path // '';
                        $route_path->{'cmd'}  = $cmd  // '';
                    }

                    $route_path->{'path'} = $under_route_path->{'path'} // '' if $route_path->{'path'} eq '';
                    $route_path->{'cmd'}  = $under_route_path->{'cmd'}  // '' if $route_path->{'cmd'}  eq '';

                    return ($route_path->{'path'}, $route_path->{'cmd'});
                  }
            };
        } else {
            ${$self->{'__LAST__'}}->{'route_path'} = $route_path;
        }
    }

    return $self;
}

sub name {
    my ($self, $name) = @_;

    my @routes_with_this_name =
      grep {$name eq ($_->{'name'} // '')} @{$self->{'__ROUTES__'}};

    throw Exception::Routes gettext('Name "%s" for route already exists', $name) if @routes_with_this_name;

    ${$self->{'__LAST__'}}->{'name'} = $name;

    return $self;
}

sub attrs {
    my ($self, @attrs) = @_;

    foreach my $attr (@attrs) {
        push(@{${$self->{'__LAST__'}}->{'attrs'}}, $attr);

        if ($attr eq 'CMD') {
            $self->type('CMD');
        } elsif ($attr eq 'FORMCMD') {
            $self->type('FORM');
            $self->process_method('_process_form');
        }
    }

    return $self;
}

sub type {
    my ($self, $type) = @_;

    throw Exception::Routes gettext('Type for this route "%s" already exists', ${$self->{'__LAST__'}}->{'route_name'})
      if defined(${$self->{'__LAST__'}}->{'type'});

    ${$self->{'__LAST__'}}->{'type'} = $type;

    return $self;
}

sub process_method {
    my ($self, $process_method) = @_;

    throw Exception::Routes gettext('Process method for this route "%s" already exists',
        ${$self->{'__LAST__'}}->{'route_name'})
      if defined(${$self->{'__LAST__'}}->{'process_method'});

    ${$self->{'__LAST__'}}->{'process_method'} = $process_method;

    return $self;
}

sub get_current_route {
    my ($self, $wi) = @_;

    $wi->{'__EXCEPTION_IN_ROUTING__'} = FALSE;

    my $method = $wi->request->method;
    my $uri    = $wi->request->uri;

    $uri =~ s/[?#][^\/]*\z//;
    $uri .= '/' if !$self->{'strictly'} && $uri !~ m/\/\z/;

    $uri = fix_utf(uri_unescape($uri));

    my @routes = $self->_get_routes_by_methods($method);

    @routes = sort {$self->_sort_routes($a, $b)} @routes;

    foreach my $route (@routes) {
        my $pattern = $route->{'pattern'};

        if (my @values = ($uri =~ m/$pattern/i)) {
            my %url_params = ();

            if (@{$route->{'params'}}) {
                throw Exception::Routes gettext('Different number of parameters')
                  unless @{$route->{'params'}} == @values;

                @url_params{@{$route->{'params'}}} = @values;
            }

            if (exists($route->{'conditions'})) {
                my $ok = TRUE;

                foreach my $condition_name (keys(%{$route->{'conditions'}})) {
                    my $check_value;
                    if (exists($url_params{$condition_name})) {
                        $check_value = $url_params{$condition_name};
                    } elsif ($wi->request->can($condition_name)) {
                        $check_value = $wi->request->$condition_name();
                    } else {
                        $check_value = $wi->request->http_header($condition_name);
                    }

                    my $condition = $route->{'conditions'}{$condition_name};

                    if (ref($condition) eq 'ARRAY') {
                        $ok = in_array($check_value, $condition);
                    } elsif (ref($condition) eq 'Regexp') {
                        $ok = $check_value =~ $condition;
                    } elsif (ref($condition) eq 'CODE') {
                        try {
                            $ok = $condition->($wi, $check_value, \%url_params);
                        }
                        catch {
                            my ($exception) = @_;

                            $wi->{'__EXCEPTION_IN_ROUTING__'} = TRUE;

                            $wi->exception_handling($exception);
                        };

                        return {} if $wi->{'__EXCEPTION_IN_ROUTING__'};
                    } else {
                        throw Exception::Routes gettext('Unknown condition type "%s"', ref($condition));
                    }

                    last unless $ok;
                }

                next unless $ok;
            }

            my $route_settings = clone($route);

            throw Exception::Routes gettext('You did not specify the path of the route "%s"', $route)
              unless defined($route_settings->{'route_path'});

            if (exists($route_settings->{'route_path'}{'handler'})) {
                $route_settings->{'handler'} = $route_settings->{'route_path'}{'handler'};
            } elsif (exists($route_settings->{'route_path'}{'controller'})) {
                try {
                    ($route_settings->{'path'}, $route_settings->{'cmd'}) =
                      $route_settings->{'route_path'}{'controller'}($wi, \%url_params);
                }
                catch {
                    my ($exception) = @_;

                    $wi->{'__EXCEPTION_IN_ROUTING__'} = TRUE;

                    $wi->exception_handling($exception);
                };

                return {} if $wi->{'__EXCEPTION_IN_ROUTING__'};
            } else {
                $route_settings->{'path'} = $route_settings->{'route_path'}{'path'};
                $route_settings->{'cmd'}  = $route_settings->{'route_path'}{'cmd'};
            }

            $route_settings->{'args'} = {map {$_ => $url_params{$_}} sort keys(%url_params)};

            return $route_settings;
        }
    }

    return {};
}

sub _get_routes_by_methods {
    my ($self, $method) = @_;

    my @routes = ();

    foreach my $route (@{$self->{'__ROUTES__'}}) {
        push(@routes, $route) if $METHODS{$method} & $route->{'methods'};
    }

    return @routes;
}

sub _sort_routes {
    my ($self, $f, $s) = @_;

    my $result = $s->{'levels'} <=> $f->{'levels'};

    if ($result == 0) {
        if (@{$f->{'params'}} && @{$s->{'params'}}) {
            $result = @{$f->{'params'}} <=> @{$s->{'params'}};
        } elsif (@{$f->{'params'}} && !@{$s->{'params'}}) {
            $result = 1;
        } elsif (!@{$f->{'params'}} && @{$s->{'params'}}) {
            $result = -1;
        }
    }

    if ($result == 0) {
        if (exists($f->{'conditions'}) && !exists($s->{'conditions'})) {
            $result = -1;
        } elsif (!exists($f->{'conditions'}) && exists($s->{'conditions'})) {
            $result = 1;
        } else {
            my $f_pattern = $f->{'pattern'};
            my $s_pattern = $s->{'pattern'};

            l(
                gettext(
                    'Warning: Routes "%s" and "%s" differences only in conditions', $f->{'route_name'},
                    $s->{'route_name'}
                )
             ) if $f->{'route_name'} =~ /$s_pattern/i && $s->{'route_name'} =~ /$f_pattern/i;
        }
    }

    return $result;
}

sub conditions {
    my ($self, %conditions) = @_;

    if (exists(${$self->{'__LAST__'}}->{'conditions'})) {
        ${$self->{'__LAST__'}}->{'conditions'} = {%{${$self->{'__LAST__'}}->{'conditions'}}, %conditions};
    } else {
        ${$self->{'__LAST__'}}->{'conditions'} = \%conditions;
    }

    return $self;
}

sub url_for {
    my ($self, $name, $params, %vars) = @_;

    my $route;
    foreach my $r (@{$self->{'__ROUTES__'}}) {
        if (($r->{'name'} // '') eq $name) {
            $route = $r;

            last;
        }
    }

    throw Exception::Routes gettex('Route with name "%s" not found', $name) unless $route;

    my $url;
    if (@{$route->{'params'}}) {
        $url = sprintf(
            $route->{'format'},
            map {uri_escape_utf8($params->{$_}) // throw Exception::Routes gettext('Expected param "%s"', $_)}
              @{$route->{'params'}}
        );
    } else {
        $url = $route->{'format'};
    }

    return $url
      . (
        %vars
        ? '?' . join('&', map {uri_escape_utf8($_) . '=' . uri_escape_utf8($vars{$_})} sort keys(%vars))
        : ''
      );
}

TRUE;
